### DISCUSSION TITLE
[BE201] 後端中階：Express 與 Sequelize- Node.js 與 MySQL 的串接影片提問

### DISCUSSION THREAD

**Discussion by powerfultraveling** [Lidemy/forum#174] [後端相關問題]
### 帳號
powerfultraveling

### 課程名稱
 Node.js 與 MySQL 的串接

### 碰到的問題
在影片 4:49 部分的地方，老師說到了這邊必須要丟一個 callback function 進去，因為沒辦法直接 return result 和 err。
![call](https://user-images.githubusercontent.com/81896228/128589809-adc209bf-0094-4e14-9fd1-62c8c2c1e317.PNG)

這個地方聽不太懂，假如像以下這樣寫的話不是就可以直接 return 了嗎?是不是我在觀念上有甚麼錯誤的地方呢?

```
const testModule = {
    getAll: () => {
        db.query("select * from users", (err, result) =>{
            if(err){
                console.log("error:" + err);
                return err;
            }else{
                console.log(result);
                return result;
            }
        })
    },

    get: (id) => {
        db.query(`select ID, username from users where ID=${id}`, (err, result, field) => {
            if(err){
                console.log("error:" + err);
                return err;
            }else{
                console.log(result);
                return result;
            }
        })
    }
}
```

### 補充說明
我了解老師用 callback function 實作的方法是在 controller 的地方再用 callback function 處理資料，但是我還是不清楚前面的觀念!
謝謝老師!

**Comment by allenliaooooooo** [comment]
沒辦法直接用 return 的意思是，即使該 function return 的值是你要的，但你在呼叫該 function 的時候，該 function 處理需要時間不會馬上回傳正確的值（非同步），而你呼叫該 function 是以同步的方式呼叫，所以程式碼執行的時候不會等到該 function 正確回傳資料回來才繼續執行，而是頭也不回的繼續往下執行了。

**Reply by allenliaooooooo** [reply]
以這段程式碼來說，`getAll` 會 return `result` 或 `err`，但因為 `db.query` 這個 function 是非同步的，所以可以想成整個 `getAll` 都是非同步的，當你在呼叫 `getAll` 的時候，例如 `const response = getAll()`，你預期 `resoponse` 會接收到 `getAll()` 執行後所 return 的 `result` 或 `err`，但因為 JavaScript 是同步執行的，所以宣告完 `response` 之後還沒拿到 `getAll()` 的 return 就往下繼續執行了，而沒有拿到 `getAll()` 回傳的 `response` 就會變成 `undefined`

```javascript
const testModule = {
    getAll: () => {
        db.query("select * from users", (err, result) =>{
            if(err){
                console.log("error:" + err);
                return err;
            }else{
                console.log(result);
                return result;
            }
        })
    },

    get: (id) => {
        db.query(`select ID, username from users where ID=${id}`, (err, result, field) => {
            if(err){
                console.log("error:" + err);
                return err;
            }else{
                console.log(result);
                return result;
            }
        })
    }
}
```

**Reply by aszx87410** [reply]
其實這樣講不太精確，並不是因為執行的時候還沒有結果所以回傳的東西才沒有結果，getAll 裡面的 return 是 db.query 的 callback 的 return，不是 getAll 的，getAll 沒有任何 return

**Reply by allenliaooooooo** [reply]
對耶！不小心忽略掉 `getAll` 其實沒有 return 了，還把 `db.query` 的 callback 的 return 當成 `getAll` 的 return QQ

**Comment by aszx87410** [comment]
你可以試試看，你真的有拿到想要的結果嗎？不是有 console 就代表有拿到，你要 const result = get() 之後 result 有東西才代表有拿到

**Reply by powerfultraveling** [reply]
剛剛打完自我解答之後才發現你們已經幫我解答了! 謝謝你們! 剛剛看完前輩們的提點後，想通了整個程式馬這樣設計的理由，謝謝你們!

後來發現如果用 callback function 的方式，我不用再管 getAll() 這個 function 實際上的值為何了，因為所有要對資料做的動作，都在 callback function 裡面完成了。解決了同步的問題又解決了資料很難從 query() 裡面拿出來的窘境。

謝謝你們!!!

**Comment by powerfultraveling** [comment]
## 自我解答(一)

### 解答
假如不用 callback function 去處理從資料庫來的資料的話，程式跑到一半有可能會為了要等待資料成功送達，而被卡在拿資料的那一行，整個程式碼就陷入被 blocking 的窘境。

### 詳解

在這一個章節，我會比較自己的方法以及老師使用 callback 的方法在實作上的差異，讓遇到同樣問題的同學可以從中聊解背後的理由:

#### 不用 callback function 的方法

假如用我一開始所提出的方法:
```
//module

const db = require("../db");// 引用連接上 mysql 的函式

const testModule = {
    getAll: () => {
        db.query("select * from users", (err, result) =>{
            if(err){
                console.log("error:" + err);
                return err;
            }else{
                console.log(result);
                return result;
            }
        })
    },

    get: (id) => {
        db.query(`select ID, username from users where ID=${id}`, (err, result, field) => {
            if(err){
                console.log("error:" + err);
                return err;
            }else{
                console.log(result);
                return result;
            }
        })
    }
}
```

我的確可以從 database 拿到正確的資料，並且直接將資料傳給 controller，但是到了 controller 去調用資料的時候就會發現問題大了:
```
//controller

const testModule = require("../module/test")//引用剛剛在 module 的檔案所寫的函式
const testControler = {
    getAll: (req, res) => {
        const data = testModule.getAll();
        res.render("tests", {
            data: data
        })
    },

    get: (req, res) => {
        const id = req.params.id;
        const data = testModule.get(id);
        res.render("test",  {
             data: data  
        })
}
```

這一段程式碼如果單以邏輯慢慢推演的話，可能不會發現問題，但是這些資料既然是 module 從 database 請求要來的資料，資料會不會有可能要很久才拿到呢?

假如因為一些因素，導致資料拿了很久才拿到，因為目前的程式是同步的跑法，所以在資料拿到之前，程式會一直卡在這，，不會往後執行，換言之，整個程式可能有很多東西要跑，但是全部要等到這一行跑完才能夠繼續跑，非常的沒效率。

為了解決這個問題，很簡單，就是把它改為非同步的方式拿取資料。


#### 使用 callback function 的方法
以下是老師所使用的方法:
```
//module

const db = require("../db");

const testModule = {
    getAll: (cb) => {
        db.query("select * from users", (err, result) => {
            if(err){
                cb(err);
                return
            }
            cb(null, result);
        })
    },

    get: (id, cb) => {
        db.query(`select ID, username from users where ID=${id}`,(err, result) => {
            if(err){
                cb(err);
                return
            }
            cb(null, result);
        })
    }
}

module.exports = testModule; 
```
老師傳了一個叫做 cb 的參數作為 callback function 進去 get()、getAll() 兩個 function 裡面。並且在 query() 裡面呼叫他來處理傳回來的值。
看到這邊能會有點無煞煞，接下來來到 controller 的檔案，一切就真相大白了!

```
const testModule = require("../module/test")
const testControler = {
    getAll: (req, res) => {
        testModule.getAll((err, result) => {
            if(err){
                return err;
            }
            res.render("tests", {
                test: result
            })
        });
    },

    get: (req, res) => {
        let id = req.params.id;
        testModule.get(id, (err, result) => {
            if(err){
                console.log(err);
                return
            }
            res.render("test", {
                data: result
            })
        });

    }
}

module.exports = testControler
```
這邊老師定義好了裡面的 callback function 怎麼運作，這麼一來，整個程式碼就不會被卡在這裡可以繼續往下跑了!

**Reply by aszx87410** [reply]
沒有喔，應該不是這樣的XD 這樣的話還是沒有理解為什麼要 callback，先講結論，結論是 callback 是必要的，因為沒有不需要 callback 也能達成的做法

原因是，db.query 這個函式本來就是非同步的，所以：

> 我的確可以從 database 拿到正確的資料，並且直接將資料傳給 controller

這一段話應該是錯的，因為你這種寫法沒有辦法把資料傳給 controller，你可以實際試試看，應該沒有辦法才對

 如果這一段是錯的，那後續推導出來的東西也都會是錯的了，在這邊會用 callback 不是因為「害怕等很久，程式會卡住」，db.query 是非同步的，所以不會卡住，而是因為「如果不用 callback，根本沒辦法正確拿到資料」

你可以試試看你說的不用 callback 的寫法，應該是拿不到正確的東西的
